### PR TITLE
Issue #7290: Add support for enhanced instanceof

### PR DESCRIPTION
**Pull request by nmancus1**
Issue #7290: Add support for "enhanced instanceof with pattern matching" grammar

References:
https://docs.oracle.com/en/java/javase/14/docs/specs/patterns-instanceof-jls.html#jls-6.3.1
https://openjdk.java.net/jeps/305
http://cr.openjdk.java.net/~gbierman/jep305/jep305-20191021/specs/patterns-instanceof-jls.html#jls-14

Java grammar per JEP:
_RelationalExpression:_
     _..._
     _RelationalExpression instanceof ReferenceType_
     _RelationalExpression instanceof Pattern_

_Pattern:_
     _ReferenceType Identifier_

Rationale:
We cannot reuse Checkstyle's `VARIABLE_DEF` node here; since a pattern variable has no modifiers, we must create a new token and node in the AST.  So, we have decided on `PATERN_VARIABLE_DEF`.

### CODE REVIEW

**Comment by nmancus1** [comment]
Check Regression Reports:

AST Regression Report:
https://checkstyle-reports-java14.s3.amazonaws.com/reports/diff_antlr_instanceof-pattern-var-def_7-5/index.html

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g [hunk 1]
```diff
@@ -1378,18 +1380,36 @@ equalityExpression
...
+        )?
         (    (options{warnWhenFollowAmbig=false;} :     (    LT^
                 |    GT^
                 |    LE^
                 |    GE^
                 )
                 shiftExpression
-            )*
-
+             )*
```
Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g:1395]: Alignment issue.

Comment (nmancus1) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g]: Fixed.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 2]
```diff
@@ -3540,6 +3540,34 @@
...
+     * if (obj instanceof String str) { }
+     * </pre>
+     * <p>parses as:</p>
+     * <pre>
+     * LITERAL_IF (if)
+     *  |--LPAREN (()
+     *  |--EXPR
+     *  |   `--LITERAL_INSTANCEOF (instanceof)
+     *  |       `--IDENT (obj)
+     *  |           `--PATTERN_VARIABLE_DEF
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:3558]: I do not understand why PATTERN appear under IDENT, it should be sibling.

```
$ cat Test.java 
public class Test {
 void foo() {
  String s = "";
  boolean b = s instanceof String;
 }
}

$ java $RUN_LOCALE -jar /var/tmp/checkstyle-8.33-all.jar -t Test.java | grep -C 3 instanceof | sed "s/    |       |   //"
|--IDENT -> b [4:10]
`--ASSIGN -> = [4:12]
    `--EXPR -> EXPR [4:16]
        `--LITERAL_INSTANCEOF -> instanceof [4:16]
            |--IDENT -> s [4:14]
            `--TYPE -> TYPE [4:27]
                `--IDENT -> String [4:27]
```

even in txt inputs you have them as sibling.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14InstanceofWithPatternMatching.java [hunk 3]
```diff
@@ -1,9 +1,13 @@
 //non-compiled with javac: Compilable with Java14
 package com.puppycrawl.tools.checkstyle.grammar.java14;
 
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.List;
 import java.util.Locale;
 
+import org.w3c.dom.ls.LSOutput;
```
Comment (pbludov) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14InstanceofWithPatternMatching.java:9]: This import is not used. What is it for?

Comment (pbludov) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14InstanceofWithPatternMatching.java:9]: `java.util.List` is also not used.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14InstanceofWithPatternMatching.java [hunk 4]
```diff
@@ -59,18 +63,154 @@ public void setModel(String model) {
...
+        else if (o1 instanceof Byte by) formatted = String.format("byte %d", by);
+        else if (o1 instanceof Long l) formatted = String.format("long %d", l);
+        else if (o1 instanceof Double d) formatted = String.format("double %f", d);
+        else if (o1 instanceof String s) formatted = String.format("String %s", s);
+        else formatted = String.format("Something else "+ o1.toString());
+
+        do {
+            L4: break L4;
+        } while (!(o1 instanceof String s));
+    }
```
Comment (pbludov) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14InstanceofWithPatternMatching.java:199]: Please add a code snippet with the full class name, something like
```java
if (o instanceof org.example.Foo foo)
```

Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14InstanceofWithPatternMatching.java:200]: Fixed.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14InstanceofWithPatternMatching.java (hunk 3 above)
Comment (nmancus1) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java14/InputJava14InstanceofWithPatternMatching.java]: Fixed.

**Pull request merged**